core/src/main/scalajvm/scalapb/zio_grpc/server/ZServerCallHandler.scala

```diff
@@ -95,6 +96,51 @@ object ZServerCallHandler {
     streamingInput(
       runtime,
       (req, requestContext, call) =>
-        impl(req).provideEnvironment(ZEnvironment(requestContext)).foreach(call.sendMessage)
+        loop(call, requestContext, impl(req).provideEnvironment(ZEnvironment(requestContext)))
     )
+
+  private def loop[Res](
+      call: ZServerCall[Res],
+      ctx: RequestContext,
+      stream: ZStream[Any, Status, Res]
+  ): ZIO[Any, Status, Unit] = {
+    // Take from the queue until either we reach the end of it or
+    // the client is not ready to receive more messages.
+    def innerLoop(queue: Dequeue[Exit[Option[Status], Res]]) =
+      queue.take
+        .flatMap {
+          case Exit.Success(elem)  =>
+            call.sendMessage(elem).as(true)
+          case Exit.Failure(cause) =>
+            cause.failureOrCause match {
+              case Left(Some(status)) => ZIO.fail(status)
+              case Left(None)         => ZIO.succeed(false)
+              case Right(e)           => ZIO.fail(e.dieOption.foldLeft(Status.INTERNAL)(_.withCause(_)))
+            }
+        }
+        .repeatWhile(_ && call.isReady)
+
+    // Acquire a permit then read as much as possible from the queue, stop
+    // when the queue has offered its last element (i.e. Exit.fail(None))
+    def outerLoop(queue: Dequeue[Exit[Option[Status], Res]]) =
+      (ctx.canSend.acquire.commit *> innerLoop(queue)).repeatWhile(identity)
+
+    ZIO
+      .scoped[Any](
+        stream
+          .catchAllCause(e =>
+            if (e.isInterrupted)
+              ZStream.failCause(e)
+            else
+              e.dieOption.fold(ZStream.failCause(e))(t => ZStream.fail(Status.INTERNAL.withCause(t)))
+          )
+          // ^ This is necessary: https://github.com/zio/zio/issues/7518
```

Comment:
Is it possible for the unit test to cover the scenario of "dying stream is turned into a queue" (https://github.com/zio/zio/issues/7518) so once the workaround is removed we know it works?

Reply:
Done. You had a good insight, as my implementation was incorrect. I decided to just take the fixed ZIO version, which can then be removed when 2.0.4 comes out.

Reply:
ZIO 2.0.4 is out, so I can rebase once this project is updated and get rid of the code I copied.